### ISSUE TITLE
Managing Vector configuration in Kubernetes environment via CRDs

### ISSUE THREAD

**Issue by MOZGIII** [timberio/vector#3018]
We can add support for managing Vector configuration via Custom Resouce Definitions when running in the Kubernetes environment.

### Motivation

- Allow separation of responsibilities - applications can ship with their own Vector config bits, prepared by the same teams that works on the service, rather than the ops team doing all the work
- Rely on native Kubernetes capabilities for access control and configuration management - fewer headaches for the ops team.
- Enable sharing a single Vector `DaemonSet` instance among multiple independent cluster users, utilizing resources more efficiently.

### Context

- This was explored briefly at the [RFC](https://github.com/timberio/vector/blob/master/rfcs/2020-04-04-2221-kubernetes-integration.md#custom-crds).
- There is the competition doing this already, for instance [banzaicloud logging operator](https://github.com/banzaicloud/logging-operator) uses fluentd.
- In the current state of things, a Vector instance is deployed into the cluster, and it has a centralized configuration, managed as part of the Vector deployment.
- There currently are ways to enable sharing this Vector instance and allowing configuration to be gathered from multiple apps, but all of those are hacky; we can offer a proper solution.

**Comment by afoninsky** [comment]
Indeed, operator-way is more native for k8s and gives more flexibility.

Few more thoughts related to a topic:
1. Various flows may belong to different vector instances. It can be useful to have "**Kind: Vector**" with specification how to deploy Vector itself (_instances count, daemonset/deployment/statefulset, antiaffinity, resources requests/limits, hpa_) and specify it in sinks/flows. It brings more flexibility in deploying, ability to control vector components more granular, use "microservice-like" approach when broken components don't not affect other flows.

2. It's good to support clean yaml instead/together with described:
```
    raw: |
      drop_invalid = true
```
=>
```
    config:
      drop_invalid: true
```
Use case: ability to use native tools to generate manifests from templates (jsonnet/tanka, pulumi, etc...)

3. In case of operators it's possible to bring maturity model which will be useful for companies. Just an example of idea: https://sdk.operatorframework.io/docs/operator-capabilities/

**Comment by binarylogic** [comment]
Thanks @afonisky. @MOZGIII, where do you think this sits in our Kubernetes roadmap phases? 2?

**Comment by MOZGIII** [comment]
This would be phase 3. We'll probably need a full-blown RFC for this.

**Comment by zvlb** [comment]
Now you can use [Kubernetes Vector Operator](https://github.com/kaasops/vector-operator) for configuring Vector in Kubernetes with CRDs

**Comment by yuriy-yarosh** [comment]
I think it would've make more sense to provide a backwards compatibility layer with [Grafana Agent Operator](https://grafana.com/docs/agent/latest/operator/architecture/) instead, and reuse the respective CRD's that had been fairly widespread.

Also it would've been really nice if it was possible to use Vector as an agent for [Grafana Faro](https://grafana.com/oss/faro/),

**Comment by spencergilbert** [comment]
> I think it would've make more sense to provide a backwards compatibility layer with [Grafana Agent Operator](https://grafana.com/docs/agent/latest/operator/architecture/) instead, and reuse the respective CRD's that had been fairly widespread.

There's probably some amount of re-use that can be done, but a quick look at those CRD's it may be difficult to translate some things. Definitely worth consideration.

> Also it would've been really nice if it was possible to use Vector as an agent for [Grafana Faro](https://grafana.com/oss/faro/),

Feel free to open a feature request for this!

**Comment by yuriy-yarosh** [comment]
> Definitely worth consideration.

@spencergilbert Yes, backwards compat, with better perf, would make Vector into a drop-in replacement for grafana agent stack.

Added #17591
